### PR TITLE
engine parallelism fix

### PR DESCRIPTION
**Pull request by Mzack9999**

### CODE REVIEW

File: v2/pkg/requests/bulk-http-request.go [hunk 1]
```diff
@@ -121,39 +115,26 @@ func (r *BulkHTTPRequest) makeHTTPRequestFromModel(baseURL string, data string,
 	return &HttpRequest{Request: request}, nil
 }
 
-func (r *BulkHTTPRequest) StartGenerator() {
-	r.generatorChan = r.generator(r.basePayloads)
+func (r *BulkHTTPRequest) CreateGenerator(URL string) {
+	if r.gsfm == nil {
```
Comment (manuelbua) [inline v2/pkg/requests/bulk-http-request.go:119]: Looks like this `r.gsfm` access should be protected as well?

Comment (Mzack9999) [reply v2/pkg/requests/bulk-http-request.go]: Good catch! I'm still moving things around, thanks for keeping an eye on this!

**Comment by bauthard** [comment]
Nuclei got **speed** booster in this PR 🔥 🔥 🔥   @Mzack9999 amazing work 👏👏

**Comment by manuelbua** [comment]
As i'm working on #149 i merged this into a parallel branch where i'm doing some testing and i found out two additional bugs.

### Duplicated URLs in input

If the supplied input file contains identical URLs then the behavior is different: the first defined request in a template will be issued two times while the remaining requests, if any, will be sent one time only. I only discovered this because i had never-completing progress bars until i tried to de-deuplicate my input.

Example input file:
```
http://localhost:9999
http://localhost:9999
```

Command:
```
go run -race . -l localhosts -t vulnerabilities/microstrategy-ssrf.yaml
```

### Race condition on `e.Results = true`?

Although this is very hard to replicate at times, to trigger this you may want to run all the `technologies/` templates at once: i'm not really sure this is a false positive or not, at times it's very much difficult to spot it, and having duplicated hosts in the input doesn't seem to alter the behavior at all; this was one of the latest traces i could get:

Command:
```
go run -race . -l localhosts -t technologies/ -timeout 2
```

```
WARNING: DATA RACE
Write at 0x00c000a10001 by goroutine 66:
  github.com/projectdiscovery/nuclei/v2/pkg/executer.(*HTTPExecuter).handleHTTP()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/pkg/executer/executer_http.go:194 +0x6a0
  github.com/projectdiscovery/nuclei/v2/pkg/executer.(*HTTPExecuter).ExecuteHTTP()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/pkg/executer/executer_http.go:116 +0x361
  github.com/projectdiscovery/nuclei/v2/internal/runner.(*Runner).processTemplateWithList.func1()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/internal/runner/runner.go:325 +0x289

Previous write at 0x00c000a10001 by goroutine 76:
  github.com/projectdiscovery/nuclei/v2/pkg/executer.(*HTTPExecuter).handleHTTP()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/pkg/executer/executer_http.go:194 +0x6a0
  github.com/projectdiscovery/nuclei/v2/pkg/executer.(*HTTPExecuter).ExecuteHTTP()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/pkg/executer/executer_http.go:116 +0x361
  github.com/projectdiscovery/nuclei/v2/internal/runner.(*Runner).processTemplateWithList.func1()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/internal/runner/runner.go:325 +0x289

Goroutine 66 (running) created at:
  github.com/projectdiscovery/nuclei/v2/internal/runner.(*Runner).processTemplateWithList()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/internal/runner/runner.go:320 +0x550
  github.com/projectdiscovery/nuclei/v2/internal/runner.(*Runner).processTemplateRequest()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/internal/runner/runner.go:256 +0x19e
  github.com/projectdiscovery/nuclei/v2/internal/runner.(*Runner).RunEnumeration.func3()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/internal/runner/runner.go:216 +0x23e

Goroutine 76 (finished) created at:
  github.com/projectdiscovery/nuclei/v2/internal/runner.(*Runner).processTemplateWithList()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/internal/runner/runner.go:320 +0x550
  github.com/projectdiscovery/nuclei/v2/internal/runner.(*Runner).processTemplateRequest()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/internal/runner/runner.go:256 +0x19e
  github.com/projectdiscovery/nuclei/v2/internal/runner.(*Runner).RunEnumeration.func3()
      /Users/manuel/dev/go/projectdiscovery/nuclei/v2/internal/runner/runner.go:216 +0x23e
==================
[INF] [bigip-config-utility] Loaded template BIG-IP Configuration Utility detected (@dwisiswant0) [informative]
Found 1 data race(s)
exit status 66

```

**Comment by Mzack9999** [comment]
@manuelbua Great work on this again, latest commits should have solved the issues. Thanks for your support!

**Pull request merged**